xmtp_mls/README.md

```diff
@@ -122,41 +116,58 @@ The [following diagram](https://app.excalidraw.com/s/4nwb0c8ork7/6pPH1kQDoj3) il
 
 Simplified high level flow for adding members to a group:
 
-1. Consume Key Packages for all new members
 1. Create a `group_intent` for adding the members
-1. Sync the state of the group with the network
-1. Convert the intent into concrete commit and welcome messages for the current epoch
+1. Consume Key Packages for all new members
+1. Convert the intent (and any pending proposals) into concrete commit and welcome messages for the current epoch
+   1. Write the welcome messages to the `post_commit_data` field for later
 1. Publish commit message
 1. Sync the state of the group with the network
 1. If no conflicts: Publish welcome messages to new members.
-   If conflicts: Go back to step 4 and try again
+   If conflicts: Go back to step 2 and try again (reset the intent's state to `TO_SEND` and clear the `data` and `post_commit_data` fields)
 
 ### Remove members from a group
 
 Simplified high level flow for removing members from a group:
 
 1. Create a `group_intent` for removing the members
-1. Sync the state of the group with the network
-1. Convert the intent into concrete commit for the current epoch
+1. Convert the intent (and any pending proposals) into concrete commit for the current epoch
+   1. If there are welcome messages, write them to the `post_commit_data` field for later
 1. Publish commit to the network
 1. Sync the state of the group with the network
-1. If no conflicts: Done.
-   If conflicts: Go back to step 3 and try again
+1. If no conflicts: Done. If there are welcome messages in the `post_commit_data` field, publish them.
+   If conflicts: Go back to step 2 and try again (reset the intent's state to `TO_SEND` and clear the `data` and `post_commit_data` fields)
 
 ### Send a message
 
 Simplified high level flow for sending a group message:
 
 1. Create a `group_intent` for sending the message
+1. Commit any pending proposals
+   1. If there are welcome messages, write them to the `post_commit_data` field for later
 1. Convert the intent into a concrete message for the current epoch
 1. Publish message to the network
 1. Sync the state of the group with the network (can be debounced or otherwise only done periodically)
-1. If no conflicts: Mark the message as committed. If conflicts: Go back to step 2.
+1. If no conflicts: Mark the message as committed. If there are welcome messages in the `post_commit_data` field, publish them.
+   If conflicts: Go back to step 2 and try again (reset the intent's state to `TO_SEND` and clear the `data` and `post_commit_data` fields)
+
+### Syncing group state
+
+The server maintains an inbound topic for each group, and a single inbound topic for the client's identity. For each topic, the client maintains a row that stores `last_synced_payload_id` and `lock_sync_until_ns` fields.
+
+1. In a single transaction, validate that `lock_until_ns` is not set to a value greater than `now()`, set it to a timeframe in the future, and fetch the `last_message_timestamp_ns`
```

Comment:
I also have this logic broken out as a separate ticket [here](https://github.com/xmtp/libxmtp/issues/292)